File: 378-kth-smallest-element-in-a-sorted-matrix/378-kth-smallest-element-in-a-sorted-matrix.py
class Solution:
    # complexity, time O(log(k)*n^2), space O(k)
    def kthSmallest(self, matrix: List[List[int]], k: int) -> int:
        h = []
        heapq.heapify(h)
        for i in range(len(matrix)):
            for j in range(len(matrix[0])):
                if len(h) < k:
                    heapq.heappush(h, -matrix[i][j])
                elif -matrix[i][j] > h[0]:
                    heapq.heappushpop(h, -matrix[i][j])
        return -heapq.heappop(h)
        
    # time limit exceeded
    # complexity, time O(n^2) (worst case O(k*n^2)), space O(k)
    """
    def kthSmallest(self, matrix: List[List[int]], k: int) -> int:
        d = deque()
        for i in range(len(matrix)):
            for j in range(len(matrix[0])):
                if len(d) < k:
                    d.append(matrix[i][j])
                    p = len(d) - 1
                    while (p > 0) and (d[p] < d[p-1]):
                        d[p], d[p-1] = d[p-1], d[p]
                        p -= 1
                elif matrix[i][j] < d[-1]:
                    d.pop()
                    d.append(matrix[i][j])
                    p = k - 1
                    while (p > 0) and (d[p] < d[p-1]):
                        d[p], d[p-1] = d[p-1], d[p]
                        p -= 1
        return d[-1]
    """

File: 378-kth-smallest-element-in-a-sorted-matrix/378-kth-smallest-element-in-a-sorted-matrix.cpp
class Solution {
public:
    int kthSmallest(vector<vector<int>>& matrix, int k) {
        vector<int> h;
        make_heap(h.begin(), h.end());
        for (int i=0; i<matrix.size(); i++) {
            for (int j=0; j<matrix[0].size(); j++) {
                if (h.size() < k) {
                    h.push_back(matrix[i][j]);
                    push_heap(h.begin(), h.end());
                }
                else {
                    if (matrix[i][j] < h.front()) {
                        pop_heap(h.begin(), h.end());
                        h.pop_back();
                        h.push_back(matrix[i][j]);
                        push_heap(h.begin(), h.end());
                    }
                }
            }
        }
        return h.front();
    }
};
